docs: Add workaround for configuration_aliases tf bug (#332)
antonbabenko/pre-commit-terraform: README.md

    **Warning:** If you use Terraform workspaces, DO NOT use this workaround ([details](https://github.com/antonbabenko/pre-commit-terraform/issues/203#issuecomment-918791847)). Wait to [`force-init`](https://github.com/antonbabenko/pre-commit-terraform/issues/224) option implementation.

> 5. `terraform_validate` in a 0.15+ module with configuration_aliases errors out

Reviewer: ```suggestion
5. `terraform_validate` in a repo with Terraform module, written using Terraform 0.15+ and which uses provider `configuration_aliases` ([Provider Aliases Within Modules](https://www.terraform.io/language/modules/develop/providers#provider-aliases-within-modules)), errors out.
```

>    When running the hook where you have configuration_aliases defined as required in the terraform block, terraform will throw an error like:

Reviewer: ```suggestion
   When running the hook against Terraform code where you have provider `configuration_aliases` defined in a `required_providers` configuration block, terraform will throw an error like:
```

   >
   >
   > Error: Provider configuration not present
   > To work with <resource> its original provider configuration at provider["registry.terraform.io/hashicorp/aws"].<provider_alias> is required, but it has been removed. This occurs when a provider configuration is removed while
   > objects created by that provider still exist in the state. Re-add the provider configuration to destroy <resource>, after which you can remove the provider configuration again.

>    This is a [known issue](https://github.com/hashicorp/terraform/issues/28490) with Terraform and how providers are initialized in 0.15 and later. To work around this for now you can add an exclude to the `terraform_validate` hook like this:

Reviewer: ```suggestion
   This is a [known issue](https://github.com/hashicorp/terraform/issues/28490) with Terraform and how providers are initialized in Terraform 0.15 and later. To work around this you can add an `exclude` parameter to the configuration of `terraform_validate` hook like this:
```

   ```yaml
   - id: terraform_validate

>      exclude: [^/].+$

Reviewer: This is a bit ambiguous regex, which means "every item, consisting of a minimum of two chars, with NOT `/` in it".
Could you please elaborate on what the real-life example would be?
Is it sort of "exclude everything"? If yes, then it should be `^.+$`, imho.

Author: Oops, that's actually wrong.  That should be `exclude: [^/]+$`.

The idea is to exclude the files in directory where the module in question is, but not the subdirectories so that the examples get processed.  So the regex should be "match any string that is 1 or more characters long that doesn't have a slash in it".  So if it's a root module like say, tf-aws-modules/terraform-aws-sqs then you'd use the above regex, and it would not process the module itself, but would validate all of the examples.

OTOH, if it's a monorepo with a lot of modules under `modules/` then you would want something like `exclude: modules/[^/]+$/[^/]+$` to exclude just the contents of each of the module directories.

Author: I fixed the regex and amended it to also explain how to handle where the module is not in the root of the repository.

   ```

>    This will exclude the root directory from being processed by this hook. Then add a subdirectory like "examples" or "tests" and put an example implementation in place that defines the providers with the proper aliases, and this will give you validation of your module through the example.

Reviewer: ```suggestion
   This will exclude the root directory from being processed by this hook. Then add a subdirectory like "examples" or "tests" and put an example implementation in place that defines the provider(s) with `configuration_aliases`, and this will give you validation of your Terraform module through the example.
```

Reviewer: Also could you please add an example of how to «add a subdirectory like "examples" or "tests"»?

Author: That could be challenging to do without making this excessively long.  Would it be sufficient, do you think, if I were to link to a module in terraform-aws-modules that has examples?  Maybe [terraform-examples-sqs](https://github.com/terraform-aws-modules/terraform-aws-sqs/tree/master/examples/complete)?

Reviewer: I was thinking of something like `include: ^examples/`

Also I didn't get what you mean about linking a module in terraform-aws-modules 🤔

Author: Oh.. that's not necessary with my exclude regex.  Since it only excludes the root path for the module, the examples under it still get run.  I thought you mean how to create an example for a module.

Reviewer: Aha, got you. Indeed... Apologies for the confusion.

   Alternately, you can use [terraform-config-inspect](https://github.com/hashicorp/terraform-config-inspect) and use a variant of [this script](https://github.com/bendrucker/terraform-configuration-aliases-action/blob/main/providers.sh) to generate a providers file at runtime:

   ```bash
   terraform-config-inspect --json . | jq -r '
     [.required_providers[].aliases]
     | flatten
     | del(.[] | select(. == null))
     | reduce .[] as $entry (
       {};
       .provider[$entry.name] //= [] | .provider[$entry.name] += [{"alias": $entry.alias}]
     )
   ' | tee aliased-providers.tf.json
   ```

>    Save this as .generate-providers.sh in the root of your repository and place it in the and add a hook to run that first, like so:

Reviewer: ```suggestion
   Save it as `.generate-providers.sh` in the root of your repository and add a `pre-commit` hook to run it before all other hooks, like so:
```

   ```yaml

>    - repo: local
>      hooks:
>        - id: generate-terraform-providers
>           name: generate-terraform-providers
>           require_serial: true
>           entry: .generate-providers.sh
>           language: script
>           files: (\.tf|\.tfvars)$
>           pass_filenames: false

Reviewer: ```suggestion
   repos:
     - repo: local
       hooks:
         - id: generate-terraform-providers
           name: generate-terraform-providers
           require_serial: true
           entry: .generate-providers.sh
           language: script
           files: \.tf(vars)?$
           pass_filenames: false

     - repo: https://github.com/pre-commit/pre-commit-hooks
[…]
```

ps: please adjust indentation in general if I messed it up

   ```

>    **Note:** This latter method will leave an "aliased-providers.tf.json" file in your path. You will either want to automate a way to clean this up or add it to your gitignore or both.

Reviewer: ```suggestion
   **Note:** The latter method will leave an "aliased-providers.tf.json" file in your repo. You will either want to automate a way to clean this up or add it to your `.gitignore` or both.
```

   Alternately, you can use [terraform-config-inspect](https://github.com/hashicorp/terraform-config-inspect) and use a variant of [this script](https://github.com/bendrucker/terraform-configuration-aliases-action/blob/main/providers.sh) to generate a providers file at runtime:

   ```bash
   terraform-config-inspect --json . | jq -r '
     [.required_providers[].aliases]
     | flatten
     | del(.[] | select(. == null))
     | reduce .[] as $entry (
       {};
       .provider[$entry.name] //= [] | .provider[$entry.name] += [{"alias": $entry.alias}]
     )
   ' | tee aliased-providers.tf.json
   ```

   Save it as `.generate-providers.sh` in the root of your repository and add a `pre-commit` hook to run it before all other hooks, like so:
   ```yaml
   - repos:
     - repo: local
       hooks:
         - id: generate-terraform-providers
            name: generate-terraform-providers
            require_serial: true
            entry: .generate-providers.sh
            language: script

>             files: (\.tf|\.tfvars)$

Reviewer: ```suggestion
            files: \.tf(vars)?$
```

            pass_filenames: false

>      - repo: https://github.com/pre-commit/pre-commit-hooks

Reviewer: ```suggestion

     - repo: https://github.com/pre-commit/pre-commit-hooks
```
